andela/haven-ah-backend
# #161342845 Create endpoints to follow and unfollow a user

#### What does this PR do?
Create endpoints to enable user follow and unfollow another user.
- follow user: **POST** `api/v1/profiles/:username/follow`.
- unfollow user: **DELETE** `api/v1/profiles/:username/follow`.

#### Description of task to be completed 
- add a `Follower` model.
- create functions to create and remove followers.
- create endpoints to follow and unfollow a user
- add a custom middleware that validates url parameters.

#### How should this be manually tested?
- on the terminal run `git clone https://github.com/andela/haven-ah-backend.git`.
- run git checkout `ft-follow-user-161342845`
- ensure that PostgreSQL is running on your local machine.
- run `npm install` to install the npm packages.
- run `npm test` to run the unit and integration tests.
To test on postman

- run `npm run start-dev` and open up postman.
- to test the routes above, the user needs to be authenticated.
- signup (`localhost:5000/api/v1/users/signup`) and use email and password to log in (`localhost:5000/api/v1/users/signin`). Copy token and pass to the `x-access-token` header to test the endpoints above.

Sample responses

if following is successful, Response body shape is
```
{
    "status": 201,
    "message": "You have followed this user."
}
```
if user tries to follow his/her self
```
{
    "status": 400,
    "message": "Sorry, you can not follow yourself. You can only follow other users."
}
```
if user is already a follower
```
{
    "status": 400,
    "message": "Sorry. You already follow this user"
}
```

For **DELETE** `api/v1/profiles/:username/follow`, sample responses are
if unfollowing is successful
```
{
    "status": 200,
    "message": "You have unfollowed this user."
}
```

if a user tries to unfollow a user they don't currently follow
```
{
    "status": 400,
    "message": "Sorry. You currently don't follow this user."
}
```
if a user tries to unfollow his/her self
```
{
    "status": 400,
    "message": "Sorry, you can not unfollow yourself."
}
```

#### Any background context you want to provide?
Modifications have been made to the getUser repository functions.


#### What are the relevant pivotal tracker stories?
#161342845

#### Screenshots (if appropriate)
<img width="1087" alt="screen shot 2018-11-05 at 5 42 16 am" src="https://user-images.githubusercontent.com/29674347/47978339-9a7b1b00-e0bd-11e8-9301-d0a3209c456b.png">

<img width="1093" alt="screen shot 2018-11-05 at 5 42 55 am" src="https://user-images.githubusercontent.com/29674347/47978355-b41c6280-e0bd-11e8-897f-2fd452856576.png">

<img width="1100" alt="screen shot 2018-11-05 at 5 43 50 am" src="https://user-images.githubusercontent.com/29674347/47978368-ceeed700-e0bd-11e8-9363-51ac028d7f2f.png">

<img width="1092" alt="screen shot 2018-11-05 at 5 50 58 am" src="https://user-images.githubusercontent.com/29674347/47978552-d1056580-e0be-11e8-8a5e-d5c5df7d0a75.png">

<img width="1086" alt="screen shot 2018-11-05 at 5 41 28 am" src="https://user-images.githubusercontent.com/29674347/47978331-8505f100-e0bd-11e8-9f3e-30fc774078db.png">
